### DISCUSSION TITLE
BusinessBase as Criteria

### DISCUSSION THREAD

**Discussion by Chicagoan2016** [MarimerLLC/csla#3707] [Questions] [answered]
We need our need criteria object to verify it's properties before we send it to the main Business object to get data from the DB.
If I recall correctly, we have talked about using Business base as criteria, that way our criteria object can have its own validation rules.
Am I going in the right direction by using Business Base as criteria for this scenario?

Do we have an example some where?

Kind Regards

**Comment by hurcane** [comment]
This is exactly how we handle complex criteria. We use BusinessBase for criteria classes used for business object lookup as well as parameters for inquiries, reports, and workflows. They aren't very complex, mainly validation of required criteria, "exists" checks, ranges (from/to), min/max of numeric values, etc.

None of these criteria classes have persistence.

**Reply by Chicagoan2016** [reply]
Thank you @hurcane.
If I understand correctly, we will implement BusinessBase as criteria similar to @rockfordlhotka 's  example in #3671 .
We will feed values to criteria object, check if it's valid and then call the main business object's Fetch method?
Regards

**Reply by hurcane** [reply, marked as answer]
The pattern you describe is correct. I'm not sure how the example you cited relates to this topic. Here's a partial example of a criteria class we use when adding serial numbers to an inventory item. Note that there is **no** data portal functionality of this class. It is used strictly to validate user input. The WinForms UI has a `PropertyChanged` event handler that checks validitiy of the critieria. Leading to a line like `AddSerialBindingSource.DataSource = DataPortal.Create<ItemSerial>(critieria);` The criteria parameter is the instance of this class.
```
using Csla;
using System;
using Csla.Rules.CommonRules;
using Csla.Rules;

namespace MyNamespace
{
    [Serializable]
    public class AddSerialNumberCriteria : BusinessBase<AddSerialNumberCriteria>
    {
        public static readonly PropertyInfo<string> ContextProperty = RegisterProperty<string>(nameof(Context));
        public string Context
        {
            get => GetProperty(ContextProperty);
            set => SetProperty(ContextProperty, value);
        }

        public static readonly PropertyInfo<string> SerialNumberProperty = RegisterProperty<string>(nameof(SerialNumber));
        public string SerialNumber
        {
            get => GetProperty(SerialNumberProperty);
            set => SetProperty(SerialNumberProperty, value);
        }

        public static readonly PropertyInfo<string> ItemIDProperty = RegisterProperty<string>(nameof(ItemID));
        public string ItemID
        {
            get => GetProperty(ItemIDProperty);
            set => SetProperty(ItemIDProperty, value);
        }

        public static readonly PropertyInfo<string> LocationIDProperty = RegisterProperty<string>(nameof(LocationID));
        public string LocationID
        {
            get => GetProperty(LocationIDProperty);
            set => SetProperty(LocationIDProperty, value);
        }

        protected override void AddBusinessRules()
        {
            base.AddBusinessRules();
            BusinessRules.AddRule(new Required(ContextProperty) { Priority = -1 });
            BusinessRules.AddRule(new ValidSerialContext(ContextProperty));

            BusinessRules.AddRule(new MaxLength(SerialNumberProperty, 100) { Priority = -1 });
            BusinessRules.AddRule(new Required(SerialNumberProperty) { Priority = -1 });
            BusinessRules.AddRule(new UnusedSerialNumber(SerialNumberProperty, ContextProperty, ItemIDProperty, LocationIDProperty));
        }
    }
}
```

**Reply by Chicagoan2016** [reply]
Since UI has knowledge of criteria (in this case), do you keep your criteria object in separate file?
Regards

**Reply by hurcane** [reply]
All of our criteria classes are in a separate project from the UI project.

**Reply by Chicagoan2016** [reply]
@hurcane I mean how we usually keep the CriteriaBase inside the main business object.

**Reply by hurcane** [reply]
I have an "it depends" answer for that. The example I provided is in a separate file. But you can also find examples in our codebase where the criteria class is nested within the business class. That's the nature of a long-lived project (25 years now) with many developers.

**Reply by Chicagoan2016** [reply]
@hurcane Thank you, appreciate your help.

**Comment by rockfordlhotka** [comment]
I love the answers here. The only thing I'll add is that, in current CSLA, you'll need a create operation method so you can initially create the object.

```c#
[Create]
private void Create() {}
```
